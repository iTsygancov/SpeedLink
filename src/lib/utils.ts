import { type ClassValue, clsx } from "clsx";
import { twMerge } from "tailwind-merge";

export function cn(...inputs: ClassValue[]) {
  return twMerge(clsx(inputs));
}

export function generateAlphanumericKeysArray() {
  const keysArray = [];

  // Generate lowercase alphabetic keys
  for (let i = 65; i <= 90; i++) {
    keysArray.push(String.fromCharCode(i));
  }

  // Generate numeric keys
  for (let i = 48; i <= 57; i++) {
    keysArray.push(String.fromCharCode(i));
  }

  return keysArray;
}
