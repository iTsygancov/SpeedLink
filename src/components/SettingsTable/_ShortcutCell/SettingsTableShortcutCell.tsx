import { Badge } from "@/components/ui/badge";
import {
  Select,
  SelectContent,
  SelectItem,
  SelectTrigger,
  SelectValue
} from "@/components/ui/select";
import { TableCell } from "@/components/ui/table";
import useOperatingSystem from "@/lib/hooks/useOS";
import { useSettingsStore } from "@/lib/store/settingsStore";
import { alphanumericKeysArray, cn } from "@/lib/utils";
import { Shortcut } from "@/types";

export type SettingsTableShortcutCellProps = {
  commands: Shortcut[];
  handleSelectValueChange: (value: string, itemIndex: number) => void;
  item: Shortcut;
  itemIndex: number;
};

const SettingsTableShortcutCell = ({
  commands,
  handleSelectValueChange,
  item,
  itemIndex
}: SettingsTableShortcutCellProps) => {
  const { settings } = useSettingsStore();
  const os = useOperatingSystem();

  const renderKey = () => {
    switch (os) {
      case "MacOS":
        return "Option";
      default:
        return "Alt";
    }
  };

  const disabledKeys = commands.map(
    (item) => item.shortcut[item.shortcut.length - 1]
  );

  return (
    <TableCell className='flex items-center gap-2 font-medium'>
      <Badge
        variant='outline'
        className={cn(
          "select-none rounded",
          item.canEdit && "bg-muted opacity-75"
        )}
      >
        {renderKey()}
      </Badge>
      {settings?.useShift && (
        <Badge
          variant='outline'
          className={cn(
            "select-none rounded",
            item.canEdit && "bg-muted opacity-75"
          )}
        >
          Shift
        </Badge>
      )}
      {item.canEdit ? (
        <Select
          key={item.id}
          onValueChange={(value) => handleSelectValueChange(value, itemIndex)}
          value={item.shortcut}
        >
          <SelectTrigger
            className='flex h-auto w-auto items-center gap-0.5 rounded border px-2.5 py-0.5 text-xs'
            icon={item.canEdit}
          >
            <SelectValue placeholder={item.shortcut || ""} />
          </SelectTrigger>
          <SelectContent className='max-h-[280px]'>
            {alphanumericKeysArray.map((el) => {
              return (
                <SelectItem
                  value={el}
                  key={el}
                  disabled={disabledKeys.includes(el)}
                >
                  {el}
                </SelectItem>
              );
            })}
          </SelectContent>
        </Select>
      ) : (
        <Badge
          variant='outline'
          className='w-8 select-none justify-center rounded'
        >
          {item.shortcut}
        </Badge>
      )}
    </TableCell>
  );
};

export default SettingsTableShortcutCell;
