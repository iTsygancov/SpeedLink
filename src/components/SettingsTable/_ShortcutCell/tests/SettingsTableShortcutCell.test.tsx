import SettingsTableShortcutCell, {
  SettingsTableShortcutCellProps
} from "../SettingsTableShortcutCell";
import { initialProps } from "./SettingsTableShortcutCell.data";
import { expectToMatchSnapshot, renderTree } from "@/test/utils";
import { describe, it } from "vitest";

const testComponentWithProps = (props: SettingsTableShortcutCellProps) => {
  expectToMatchSnapshot(renderTree(<SettingsTableShortcutCell {...props} />));
};

describe("SettingsTableShortcutCell", () => {
  it("should render with empty shortcut badge", () => {
    testComponentWithProps(initialProps);
  });

  it("should render with shortcut filled shortcut badge", () => {
    testComponentWithProps({
      ...initialProps,
      item: {
        ...initialProps.item,
        shortcut: "a"
      }
    });
  });
});
