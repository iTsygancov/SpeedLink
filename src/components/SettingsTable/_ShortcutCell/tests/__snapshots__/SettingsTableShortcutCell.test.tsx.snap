// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`SettingsTableShortcutCell > should render with empty shortcut badge 1`] = `
<td
  className="p-4 align-middle [&:has([role=checkbox])]:pr-0 flex items-center gap-2 font-medium"
>
  <div
    className="inline-flex items-center border px-2.5 py-0.5 text-xs font-semibold transition-colors focus:outline-none focus:ring-2 focus:ring-ring focus:ring-offset-2 text-foreground select-none rounded"
  >
    Alt
  </div>
  <div
    className="inline-flex items-center border px-2.5 py-0.5 text-xs font-semibold transition-colors focus:outline-none focus:ring-2 focus:ring-ring focus:ring-offset-2 text-foreground select-none rounded"
  >
    Shift
  </div>
  <div
    className="inline-flex items-center border px-2.5 py-0.5 text-xs font-semibold transition-colors focus:outline-none focus:ring-2 focus:ring-ring focus:ring-offset-2 text-foreground w-8 select-none justify-center rounded"
  />
</td>
`;

exports[`SettingsTableShortcutCell > should render with shortcut filled shortcut badge 1`] = `
<td
  className="p-4 align-middle [&:has([role=checkbox])]:pr-0 flex items-center gap-2 font-medium"
>
  <div
    className="inline-flex items-center border px-2.5 py-0.5 text-xs font-semibold transition-colors focus:outline-none focus:ring-2 focus:ring-ring focus:ring-offset-2 text-foreground select-none rounded"
  >
    Alt
  </div>
  <div
    className="inline-flex items-center border px-2.5 py-0.5 text-xs font-semibold transition-colors focus:outline-none focus:ring-2 focus:ring-ring focus:ring-offset-2 text-foreground select-none rounded"
  >
    Shift
  </div>
  <div
    className="inline-flex items-center border px-2.5 py-0.5 text-xs font-semibold transition-colors focus:outline-none focus:ring-2 focus:ring-ring focus:ring-offset-2 text-foreground w-8 select-none justify-center rounded"
  >
    a
  </div>
</td>
`;
